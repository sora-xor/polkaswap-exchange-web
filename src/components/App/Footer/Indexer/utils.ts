import { IndexerType } from '@soramitsu/soraneo-wallet-web/lib/consts';

export function getIndexerName(type: IndexerType): string {
  return type.charAt(0).toUpperCase() + type.slice(1).toLowerCase();
}
