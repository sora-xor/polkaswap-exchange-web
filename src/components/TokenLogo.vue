<template>
  <span :class="tokenClasses" />
</template>

<script lang="ts">
import { Component, Mixins, Prop } from 'vue-property-decorator'
import { KnownSymbols, Asset, AccountAsset } from '@sora-substrate/util'

import TranslationMixin from '@/components/mixins/TranslationMixin'
import { LogoSize } from '@/consts'

@Component
export default class TokenLogo extends Mixins(TranslationMixin) {
  @Prop({ type: Object, default: () => null }) readonly token!: AccountAsset | Asset | any
  @Prop({ type: String, default: '' }) readonly tokenSymbol!: string
  @Prop({ type: String, default: LogoSize.MEDIUM, required: false }) readonly size!: LogoSize

  get tokenClasses (): string {
    const tokenLogoClass = 'token-logo'
    const classes = [tokenLogoClass]

    if (this.tokenSymbol) {
      classes.push(`${tokenLogoClass}--${(this.tokenSymbol).toLowerCase()}`)
    } else if (this.token && !!KnownSymbols[this.token.symbol]) {
      classes.push(`${tokenLogoClass}--${(this.token.symbol).toLowerCase()}`)
    }

    classes.push(`${tokenLogoClass}--${this.size.toLowerCase()}`)
    return classes.join(' ')
  }
}
</script>

<style lang="scss" scoped>
// TODO: Check assets list + logo titles
$tokens-list: "bridge-item-xor", "bridge-item-eth", "pswap", "val", "xor", "eth";
$token-background-color: var(--s-color-base-on-accent);

.token-logo {
  background-color: $token-background-color;
  background-image: url("~@/assets/img/token-logo-default.svg");
  background-size: 60%;
  background-repeat: no-repeat;
  background-position: 50%;
  border: 1px solid $token-background-color;
  border-radius: 50%;
  box-shadow: var(--s-shadow-tooltip);
  @each $token in $tokens-list {
    &--#{$token} {
      background-size: 100%;
    }
    &--#{$token} {
      background-image: url("~@/assets/img/#{$token}.svg");
    }
  }
}

@include element-size('token-logo--mini', 16px);
@include element-size('token-logo--small', 24px);
@include element-size('token-logo--medium');
</style>
