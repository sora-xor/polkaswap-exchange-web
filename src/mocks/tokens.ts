export const tokens = [{
  name: 'Sora',
  symbol: 'XOR',
  address: '1f9840a85d5af5bf1d1762f925bdaddc4201f984',
  balance: 10000,
  price: 55.10,
  priceChange: 12
},
{
  name: 'Kusama',
  symbol: 'KSM',
  address: '34916349d43f65bccca11ff53a8e0382a1a594a7',
  balance: 0,
  price: 0.0055,
  priceChange: 12

},
{
  name: 'Kusama',
  symbol: 'KSM1',
  address: '34916349d43f65bccca11ff53a8e0382a1a594a7',
  balance: 0,
  price: 0.0055,
  priceChange: 12

},
{
  name: 'Kusama',
  symbol: 'KSM2',
  address: '34916349d43f65bccca11ff53a8e0382a1a594a7',
  balance: 0,
  price: 0.0055,
  priceChange: 12

},
{
  name: 'Kusama',
  symbol: 'KSM3',
  address: '34916349d43f65bccca11ff53a8e0382a1a594a7',
  balance: 0,
  price: 0.0055,
  priceChange: 12

},
{
  name: 'Kusama',
  symbol: 'KSM4',
  address: '34916349d43f65bccca11ff53a8e0382a1a594a7',
  balance: 0,
  price: 0.0055,
  priceChange: 12

},
{
  name: 'Kusama',
  symbol: 'KSM5',
  address: '34916349d43f65bccca11ff53a8e0382a1a594a7',
  balance: 0,
  price: 0.0055,
  priceChange: 12

},
{
  name: 'Kusama',
  symbol: 'KSM6',
  address: '34916349d43f65bccca11ff53a8e0382a1a594a7',
  balance: 0,
  price: 0.0055,
  priceChange: 12

},
{
  name: 'Kusama',
  symbol: 'KSM7',
  address: '34916349d43f65bccca11ff53a8e0382a1a594a7',
  balance: 0,
  price: 0.0055,
  priceChange: 12

},
{
  name: 'Ether',
  symbol: 'ETH',
  address: '8adaca8ea8192656a15c88797e04c8771c4576b3',
  balance: 20000.45,
  price: 0.0099,
  priceChange: -12
}]
