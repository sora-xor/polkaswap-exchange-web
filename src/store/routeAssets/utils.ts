import { FPNumber, Operation } from '@sora-substrate/util/build';
import { Asset, AccountAsset } from '@sora-substrate/util/build/assets/types';
import { api } from '@soramitsu/soraneo-wallet-web';
import { FiatPriceObject } from '@soramitsu/soraneo-wallet-web/lib/services/subquery/types';

import { Recipient } from './types';

import type { WhitelistArrayItem } from '@sora-substrate/util/build/assets/types';

export default {
  validate(recipient: Recipient) {
    return (
      this.wallet(recipient.wallet) &&
      this.asset(recipient.asset) &&
      this.usd(recipient.usd) &&
      this.amount(recipient.amount)
    );
  },

  name(name: Nullable<string>) {
    return name;
  },

  wallet(wallet: string) {
    return api.validateAddress(wallet);
  },

  asset(asset: Asset | AccountAsset) {
    return asset;
  },

  usd(usd: Nullable<number>) {
    return usd && !isNaN(usd);
  },

  amount(amount: Nullable<number>) {
    return amount && !isNaN(amount) && Number.isFinite(amount);
  },
};

export function getTokenEquivalent(
  priceObject: FiatPriceObject,
  asset: Asset | AccountAsset | WhitelistArrayItem,
  usd: number | string
): FPNumber {
  return new FPNumber(usd).div(FPNumber.fromCodecValue(priceObject[asset.address], asset.decimals));
}

export function getAssetUSDPrice(asset, fiatPriceObject) {
  if (!asset) return null;
  return FPNumber.fromCodecValue(fiatPriceObject[asset.address], asset.decimals);
}
