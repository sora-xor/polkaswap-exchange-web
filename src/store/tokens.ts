import map from 'lodash/fp/map'
import flatMap from 'lodash/fp/flatMap'
import fromPairs from 'lodash/fp/fromPairs'
import flow from 'lodash/fp/flow'
import tokensApi from '@/api/tokens'
import { Token } from '@/types'

const types = flow(
  flatMap(x => [x + '_REQUEST', x + '_SUCCESS', x + '_FAILURE']),
  map(x => [x, x]),
  fromPairs
)([
  'GET_TOKENS_LIST'
])

function initialState () {
  return {
    tokens: []
  }
}

const state = initialState()

const getters = {
  tokens (state) {
    return state.tokens
  }
}

const mutations = {
  [types.GET_TOKENS_LIST_REQUEST] (state) {
    state.tokens = null
  },

  [types.GET_TOKENS_LIST_SUCCESS] (state, tokens: Array<Token>) {
    state.tokens = tokens
  },

  [types.GET_TOKENS_LIST_FAILURE] (state) {
    state.tokens = null
  }
}

const actions = {
  async getTokens ({ commit }) {
    commit(types.GET_TOKENS_LIST_REQUEST)
    try {
      const tokens = await tokensApi.getTokens() as Array<Token>
      commit(types.GET_TOKENS_LIST_SUCCESS, tokens)
    } catch (error) {
      commit(types.GET_TOKENS_LIST_FAILURE, error)
    }
  }
}

export default {
  types,
  state,
  getters,
  mutations,
  actions
}
