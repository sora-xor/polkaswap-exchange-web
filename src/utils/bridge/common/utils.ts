import { Operation, isEthOperation, isEvmOperation, isSubstrateOperation } from '@sora-substrate/util';
import { api as soraApi } from '@soramitsu/soraneo-wallet-web';
import { ethers } from 'ethers';

import { isUnsignedTx as isUnsignedEthTx } from '@/utils/bridge/eth/utils';
import { isUnsignedTx as isUnsignedEvmTx } from '@/utils/bridge/evm/utils';
import { isUnsignedTx as isUnsignedSubTx } from '@/utils/bridge/sub/utils';
import ethersUtil from '@/utils/ethers-util';

import type { ApiPromise } from '@polkadot/api';
import type { IBridgeTransaction } from '@sora-substrate/util';
import type { EthHistory } from '@sora-substrate/util/build/bridgeProxy/eth/types';
import type { EvmHistory } from '@sora-substrate/util/build/bridgeProxy/evm/types';
import type { SubHistory } from '@sora-substrate/util/build/bridgeProxy/sub/types';

export const waitForEvmTransactionMined = async (
  tx: ethers.TransactionResponse | null,
  replaceCallback?: (tx: ethers.TransactionResponse | null) => void
): Promise<ethers.TransactionReceipt | null> => {
  if (!tx) throw new Error('[waitForEvmTransactionMined]: tx cannot be empty!');

  try {
    const startBlock = tx.blockNumber ?? (await ethersUtil.getBlockNumber());
    const replaceableTx = tx.replaceableTransaction(startBlock);
    const txReceipt = await replaceableTx.wait();

    return txReceipt;
  } catch (error) {
    if (ethers.isError(error, 'TRANSACTION_REPLACED')) {
      const replacedTx = error.replacement;

      replaceCallback?.(replacedTx);

      return await waitForEvmTransactionMined(replacedTx, replaceCallback);
    }

    throw error;
  }
};

export const getEvmTransactionRecieptByHash = async (
  transactionHash: string
): Promise<{ fee: string; blockHash: string; blockNumber: number; from: string } | null> => {
  try {
    const receipt = await ethersUtil.getEvmTransactionReceipt(transactionHash);

    if (!receipt) throw new Error(`Transaction receipt "${transactionHash}" not found`);

    const { fee, from, blockNumber, blockHash } = receipt;

    return { fee: fee.toString(), blockHash, blockNumber, from };
  } catch (error) {
    return null;
  }
};

export const getBlockEventsByTxIndex = async (blockHash: string, index: number, api: ApiPromise) => {
  const blockEvents = await soraApi.system.getBlockEvents(blockHash, api);
  const transactionEvents = blockEvents.filter(
    ({ phase }) => phase.isApplyExtrinsic && phase.asApplyExtrinsic.toNumber() === index
  );

  return transactionEvents;
};

export const getTransactionEvents = async (blockHash: string, transactionHash: string, api: ApiPromise) => {
  const extrinsics = await soraApi.system.getExtrinsicsFromBlock(blockHash, api);
  const extrinsicIndex = extrinsics.findIndex((ext) => ext.hash.toString() === transactionHash);

  if (extrinsicIndex === -1) throw new Error(`Unable to find extrinsic "${transactionHash}" in block "${blockHash}"`);

  const transactionEvents = await getBlockEventsByTxIndex(blockHash, extrinsicIndex, api);

  return transactionEvents;
};

export const isOutgoingTransaction = (transaction: Nullable<IBridgeTransaction>): boolean => {
  if (!transaction?.type) return false;

  return [Operation.EthBridgeOutgoing, Operation.EvmOutgoing, Operation.SubstrateOutgoing].includes(transaction.type);
};

export const isUnsignedTx = (transaction: Nullable<IBridgeTransaction>): boolean => {
  if (!transaction?.type) return true;

  if (isEthOperation(transaction.type)) return isUnsignedEthTx(transaction as EthHistory);
  if (isEvmOperation(transaction.type)) return isUnsignedEvmTx(transaction as EvmHistory);
  if (isSubstrateOperation(transaction.type)) return isUnsignedSubTx(transaction as SubHistory);

  return true;
};
