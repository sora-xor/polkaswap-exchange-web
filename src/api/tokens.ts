import { tokens } from '@/mocks/tokens'

const getTokens = () => {
  return Promise.resolve(tokens)
}

export default {
  getTokens
}
