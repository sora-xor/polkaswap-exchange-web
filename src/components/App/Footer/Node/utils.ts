import { countryCodeEmoji } from 'country-code-emoji';

import store from '@/store';

export function formatLocation(code: string): Nullable<{ flag: string; name?: string }> {
  try {
    const isoCode = code.toUpperCase();
    const flag = countryCodeEmoji(isoCode);
    const location = { flag, name: '' };
    const displayRegions = store.state.settings.displayRegions;
    if (!displayRegions) {
      return location;
    }
    const name = displayRegions.of(isoCode);
    if (name) {
      location.name = name;
    }
    return location;
  } catch (error) {
    console.warn('Unsupported format of location in env.json', error);
    return null;
  }
}
