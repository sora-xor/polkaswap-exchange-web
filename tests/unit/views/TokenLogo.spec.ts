import { shallowMount, createLocalVue } from '@vue/test-utils'
import SoramitsuElements from '@soramitsu/soramitsu-js-ui'
import { tokens } from '@/mocks/tokens'

import TokenLogo from '@/components/TokenLogo.vue'
import { TranslationMock } from '../../utils'

const localVue = createLocalVue()
localVue.use(SoramitsuElements)

describe('TokenLogo.vue', () => {
  beforeEach(() => {
    TranslationMock(TokenLogo)
  })

  it('should renders correctly', () => {
    const wrapper = shallowMount(TokenLogo, {
      localVue,
      propsData: {
        token: tokens[0].symbol
      }
    })
    expect(wrapper.element).toMatchSnapshot()
  })
})
