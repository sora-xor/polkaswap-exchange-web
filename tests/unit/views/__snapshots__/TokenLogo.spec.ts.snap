// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TokenLogo.vue should renders correctly 1`] = `
<span
  class="token-logo token-logo--xor token-logo--medium"
/>
`;
